### PR TITLE
rgw/qa: Run s3tests on dbstore backend in teuthology

### PR DESCRIPTION
**Pull request by soumyakoduri**
Add a new test-suite to run s3tests on dbstore backend. 

Also compiled a list of testcases failing on dbstore currently and marked them as 'fails_on_dbstore' to be filtered in teuthology run (https://github.com/ceph/s3-tests/pull/441)

Fixes: https://tracker.ceph.com/issues/55417
Signed-off-by: Soumya Koduri <skoduri@redhat.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/common/options/rgw.yaml.in [hunk 1]
```diff
@@ -3240,6 +3240,20 @@ options:
   - rados
   - dbstore
   - motr
+- name: dbstore_db_path
+  type: str
+  level: advanced
+  desc: experimental Option to set db backend store file path
```
Comment (mattbenjamin) [inline src/common/options/rgw.yaml.in:3246]: why experimental?  this seems like a fairly plausible option I'd look for

File: src/common/options/rgw.yaml.in [hunk 2]
```diff
@@ -3240,6 +3240,20 @@ options:
...
+  default: /var/run/ceph
+  services:
+  - rgw
+- name: dbstore_db_name_prefix
+  type: str
+  level: advanced
+  desc: experimental Option to set db backend store file name prefix
```
Comment (mattbenjamin) [inline src/common/options/rgw.yaml.in:3253]: ditto

Comment (mattbenjamin) [reply src/common/options/rgw.yaml.in:3253]: (at least, I don't see what's experimental about this, even if we got rid of it at some point)

File: src/rgw/store/dbstore/README.md [hunk 3]
```diff
@@ -26,7 +26,12 @@ Restart vstart cluster or just RGW server
 
 The above configuration brings up RGW server on dbstore and creates testid user to be used for s3 operations.
 
-By default, dbstore creates .db file named 'default_ns.db' where in the data is stored.
+
+By default, dbstore creates .db file *'/var/run/ceph/dbstore-default_ns.db'* to store the data. This can be configured using below options in ceph.conf
```
Comment (mattbenjamin) [inline src/rgw/store/dbstore/README.md:30]: I do think this is much more robust, and typical of database apps

**Review by mattbenjamin — COMMENTED** [review]
I haven't run it, but it looks sane to me

File: src/common/options/rgw.yaml.in (hunk 1 above)
Comment (soumyakoduri) [reply src/common/options/rgw.yaml.in:3246]: yes..overlooked that part in copy/paste. will fix it.

File: src/common/options/rgw.yaml.in (hunk 2 above)
Comment (soumyakoduri) [reply src/common/options/rgw.yaml.in:3253]: done.

File: qa/suites/rgw/dbstore/tasks/rgw_s3tests.yaml [hunk 4]
```diff
@@ -0,0 +1,23 @@
+tasks:
+- install:
+- ceph:
+- rgw: [client.0]
+- s3tests:
+    client.0:
+      dbstore_tests: True
+      force-repo: https://github.com/soumyakoduri/s3-tests.git
```
Comment (soumyakoduri) [inline qa/suites/rgw/dbstore/tasks/rgw_s3tests.yaml:8]: This will be set  back to default once https://github.com/ceph/s3-tests/pull/441 gets merged.

**Review by dang — APPROVED** [review]

**Comment by soumyakoduri** [comment]
jenkins test make check

**Comment by soumyakoduri** [comment]
http://pulpito.front.sepia.ceph.com/soumyakoduri-2022-04-25_06:22:48-rgw:dbstore-wip-skoduri-dbstore-tests-distro-basic-smithi

File: qa/suites/rgw/dbstore/cluster.yaml [hunk 5]
```diff
@@ -0,0 +1,10 @@
+roles:
+- [mon.a, osd.0, osd.1, osd.2, mgr.0, client.0]
+openstack:
```
Comment (alimaredia) [inline qa/suites/rgw/dbstore/cluster.yaml:3]: I'm curious, what's the point of having the openstack task here?

File: qa/suites/rgw/dbstore/cluster.yaml [hunk 6]
```diff
@@ -0,0 +1,10 @@
...
+- volumes: # attached to each instance
+    count: 1
+    size: 10 # GB
+overrides:
```
Comment (alimaredia) [inline qa/suites/rgw/dbstore/cluster.yaml:7]: Should this override go into `qa/suites/rgw/dbstore/overrides.yaml`?

File: qa/suites/rgw/dbstore/tasks/rgw_s3tests.yaml [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+      git_remote: https://github.com/soumyakoduri/
+      force-branch: dbstore-tests
+      rgw_server: client.0
+      storage classes: LUKEWARM, FROZEN
+      extra_attrs: ["!fails_on_dbstore"]
+overrides:
```
Comment (alimaredia) [inline qa/suites/rgw/dbstore/tasks/rgw_s3tests.yaml:14]: Same with these overrides. Is providing overrides in multiple places how we do it in other parts of the rgw suite?

File: qa/suites/rgw/dbstore/cluster.yaml (hunk 5 above)
Comment (soumyakoduri) [reply qa/suites/rgw/dbstore/cluster.yaml:3]: I had referred to  couple of other test-suites. Will fix this and below comments.

File: qa/tasks/s3tests.py [hunk 8]
```diff
@@ -260,6 +274,7 @@ def create_users(ctx, config):
     try:
         yield
     finally:
+       if not ctx.dbstore_variable:
```
[the remark this answers is not in the document]
Comment (soumyakoduri) [reply qa/tasks/s3tests.py]: Fixed it.

File: src/rgw/store/dbstore/README.md [hunk 9]
```diff
@@ -26,7 +26,12 @@ Restart vstart cluster or just RGW server
...
+
+    [client]
+        dbstore db path = <path-to-db-file>
```
[the remark this answers is not in the document]
Comment (soumyakoduri) [reply src/rgw/store/dbstore/README.md]: done.

**Review by cbodley — APPROVED** [review]

File: qa/tasks/s3tests.py [hunk 10]
```diff
@@ -201,6 +201,20 @@ def create_users(ctx, config):
                         ],
                     )
 
+    elif ctx.dbstore_variable: 
+        # till API available to create new users on dbstore, use testid as default
+        users = {'s3 main': 'testid', 's3 alt': 'testid', 's3 tenant': 'testid'}
```
[the remark this answers is not in the document]
Comment (soumyakoduri) [reply qa/tasks/s3tests.py:206]: Have submitted https://github.com/ceph/ceph/pull/46126 to support user creation on dbstore via `radosgw-amin`. Will update this PR once that gets merged.

**Comment by soumyakoduri** [comment]
jenkins test windows

**Comment by soumyakoduri** [comment]
@dang @cbodley @alimaredia ... with https://github.com/ceph/ceph/pull/46126 merged , have updated the PR to allow users creation even on 'dbstore' (via radosgw-admin  cmd) .  Please let me know if you have any further comments.

latest results:
http://pulpito.front.sepia.ceph.com/soumyakoduri-2022-05-16_18:12:38-rgw:dbstore-wip-skoduri-dbstore-vstart-distro-basic-smithi
http://pulpito.front.sepia.ceph.com/soumyakoduri-2022-05-17_09:37:38-rgw-wip-skoduri-dbstore-vstart-distro-basic-smithi/

**Comment by soumyakoduri** [comment]
jenkins test api
jenkins test windows

**Comment by soumyakoduri** [comment]
jenkins test api

**Comment by soumyakoduri** [comment]
jenkins test windows

**Comment by soumyakoduri** [comment]
http://pulpito.front.sepia.ceph.com/soumyakoduri-2022-05-21_18:27:52-rgw-wip-skoduri-dbstore-tests-distro-basic-smithi/

**Pull request merged**